comp/core/secrets/secretsimpl/secrets.go

```diff
@@ -199,17 +210,61 @@ func isEnc(str string) (bool, string) {
 // Resolve replaces all encrypted secrets in data by executing "secret_backend_command" once if all secrets aren't
 // present in the cache.
 func (r *secretResolver) Resolve(data []byte, origin string) ([]byte, error) {
-	return r.resolve(data, origin, nil)
+	return r.resolve(data, origin, nil, nil)
 }
 
-// ResolveWithCallback resolves the secrets in the given yaml data calling the callback with the YAML path of
+// RegisterResolveCallback resolves the secrets in the given yaml data calling the callback with the YAML path of
 // the secret handle and its value
-func (r *secretResolver) ResolveWithCallback(data []byte, origin string, cb secrets.ResolveCallback) error {
-	_, err := r.resolve(data, origin, cb)
+func (r *secretResolver) RegisterResolveCallback(data []byte, origin string, cb secrets.ResolveCallback) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	// enable refreshing data by storing required parameters
+	r.refreshData = data
+	r.refreshOrigin = origin
+	r.refreshCallback = cb
+	if r.refreshInterval > 0 && r.lastRefresh.IsZero() {
+		r.lastRefresh = time.Now()
+		go func() {
+			for {
+				var since time.Duration
+				func() {
+					r.lock.Lock()
+					defer r.lock.Unlock()
+					since = time.Since(r.lastRefresh)
+				}()
+				if since.Seconds() > float64(r.refreshInterval) {
+					if err := r.Refresh(); err != nil {
+						log.Info(err)
+					}
+					since = 0
+				}
+				time.Sleep(time.Duration(r.refreshInterval) - since)
+			}
+		}()
+	}
```

Comment:
I don't know if we should worry about this with Go, but would it make sense to stop this goroutine when the agent stops? If that is the case, we could use the fx lifecycle events for that 😄

Reply:
I'm not sure if fx lifecycle would improve this, I don't have much experience with their use just yet. As is now, this routine is only sometimes started (if the `secret_refresh_interval` flag is non-zero) and will shut down fine when exiting the agent process.